File: ai/converter.py
import argparse
import os
import sys
import logging
import torch
import tensorflow as tf
from tensorflow.keras.models import load_model
import h5py
import numpy as np
from abc import ABC, abstractmethod

# Configure the logger
logger = logging.getLogger('ModelConverter')
logger.setLevel(logging.DEBUG)  # Set to DEBUG for detailed logs

# Create console handler with a higher log level
ch = logging.StreamHandler()
ch.setLevel(logging.INFO)  # Set to INFO or DEBUG as needed

# Create formatter and add it to the handlers
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
ch.setFormatter(formatter)

# Add the handlers to the logger
logger.addHandler(ch)

# Abstract Base Class for Model Converters
class ModelConverter(ABC):
    def __init__(self, input_path, output_path):
        self.input_path = input_path
        self.output_path = output_path

    @abstractmethod
    def load_model(self):
        pass

    @abstractmethod
    def convert_to_ggml(self):
        pass

    def save_ggml_model(self, ggml_model):
        """
        Saves the GGML model to the specified output path.
        """
        try:
            logger.info(f"Saving GGML model to: {self.output_path}")
            # Placeholder for actual GGML saving logic
            # Replace this with actual implementation
            with open(self.output_path, "wb") as f:
                torch.save(ggml_model, f)
            logger.info("Conversion complete.")
        except Exception as e:
            logger.error(f"Failed to save GGML model: {e}")
            raise

    def convert(self):
        """
        Full conversion process: load, convert, and save.
        """
        self.load_model()
        ggml_model = self.convert_to_ggml()
        self.save_ggml_model(ggml_model)

# PyTorch Converter
class PyTorchConverter(ModelConverter):
    def load_model(self):
        """
        Loads a PyTorch .pth model.
        """
        logger.info(f"Loading PyTorch model from: {self.input_path}")
        try:
            self.model = torch.load(self.input_path, map_location=torch.device("cpu"))
            logger.debug("PyTorch model loaded successfully.")
        except Exception as e:
            logger.error(f"Failed to load PyTorch model: {e}")
            raise

    def convert_to_ggml(self):
        """
        Converts the PyTorch model to a GGML-compatible format.
        """
        logger.info("Converting PyTorch model to GGML format...")
        try:
            ggml_model = {}
            for name, param in self.model.items():
                if isinstance(param, torch.Tensor):
                    ggml_model[name] = param.cpu().numpy()
                    logger.debug(f"Converted parameter: {name}")
            logger.debug("PyTorch model converted to GGML format successfully.")
            return ggml_model
        except Exception as e:
            logger.error(f"Failed to convert PyTorch model to GGML: {e}")
            raise

# TensorFlow Converter
class TensorFlowConverter(ModelConverter):
    def load_model(self):
        """
        Loads a TensorFlow .pb or Keras .h5 model.
        """
        logger.info(f"Loading TensorFlow model from: {self.input_path}")
        try:
            if self.input_path.endswith('.h5'):
                self.model = load_model(self.input_path)
                logger.debug("Keras model loaded successfully.")
            elif self.input_path.endswith('.pb'):
                # Placeholder for TensorFlow .pb loading logic
                # TensorFlow's SavedModel format should be used for better compatibility
                self.model = tf.saved_model.load(self.input_path)
                logger.debug("TensorFlow SavedModel loaded successfully.")
            else:
                raise ValueError("Unsupported TensorFlow model format. Use .h5 or .pb.")
        except Exception as e:
            logger.error(f"Failed to load TensorFlow model: {e}")
            raise

    def convert_to_ggml(self):
        """
        Converts the TensorFlow model to a GGML-compatible format.
        """
        logger.info("Converting TensorFlow model to GGML format...")
        try:
            ggml_model = {}
            if isinstance(self.model, tf.keras.Model):
                for layer in self.model.layers:
                    for weight in layer.weights:
                        name = weight.name
                        value = weight.numpy()
                        ggml_model[name] = value
                        logger.debug(f"Converted weight: {name}")
            else:
                # Handle other TensorFlow model types if necessary
                logger.warning("TensorFlow model type not fully supported for conversion.")
            logger.debug("TensorFlow model converted to GGML format successfully.")
            return ggml_model
        except Exception as e:
            logger.error(f"Failed to convert TensorFlow model to GGML: {e}")
            raise

# Keras Converter (if separate from TensorFlow)
class KerasConverter(ModelConverter):
    def load_model(self):
        """
        Loads a Keras .h5 model.
        """
        logger.info(f"Loading Keras model from: {self.input_path}")
        try:
            self.model = load_model(self.input_path)
            logger.debug("Keras model loaded successfully.")
        except Exception as e:
            logger.error(f"Failed to load Keras model: {e}")
            raise

    def convert_to_ggml(self):
        """
        Converts the Keras model to a GGML-compatible format.
        """
        logger.info("Converting Keras model to GGML format...")
        try:
            ggml_model = {}
            for layer in self.model.layers:
                for weight in layer.weights:
                    name = weight.name
                    value = weight.numpy()
                    ggml_model[name] = value
                    logger.debug(f"Converted weight: {name}")
            logger.debug("Keras model converted to GGML format successfully.")
            return ggml_model
        except Exception as e:
            logger.error(f"Failed to convert Keras model to GGML: {e}")
            raise

# GGML Converter Registry
class GGMLConverterRegistry:
    converters = {
        '.pth': PyTorchConverter,
        '.pt': PyTorchConverter,
        '.h5': KerasConverter,
        '.pb': TensorFlowConverter,
        # Add more mappings as needed
    }

    @classmethod
    def get_converter(cls, input_path, output_path):
        ext = os.path.splitext(input_path)[1].lower()
        converter_class = cls.converters.get(ext)
        if not converter_class:
            logger.error(f"No converter available for the file extension: {ext}")
            raise ValueError(f"No converter available for the file extension: {ext}")
        return converter_class(input_path, output_path)

def convert_model(input_path, output_path):
    """
    Determines the appropriate converter and performs the conversion.
    """
    logger.info(f"Starting conversion: {input_path} -> {output_path}")
    try:
        converter = GGMLConverterRegistry.get_converter(input_path, output_path)
        converter.convert()
    except Exception as e:
        logger.error(f"Conversion failed: {e}")
        sys.exit(1)
    logger.info("Model conversion completed successfully.")

def main():
    parser = argparse.ArgumentParser(description="Convert various AI model formats to GGML-compatible format.")
    parser.add_argument("input", type=str, help="Path to the input model file (e.g., .pth, .h5, .pb).")
    parser.add_argument("output", type=str, help="Path to save the converted GGML model (e.g., .ggml).")
    parser.add_argument("--overwrite", action="store_true", help="Overwrite the output file if it already exists.")
    parser.add_argument("--log-level", type=str, default="INFO", choices=["DEBUG", "INFO", "WARNING", "ERROR"],
                        help="Set the logging level.")

    args = parser.parse_args()

    # Configure logging level based on user input
    numeric_level = getattr(logging, args.log_level.upper(), None)
    if not isinstance(numeric_level, int):
        logger.error(f"Invalid log level: {args.log_level}")
        sys.exit(1)
    ch.setLevel(numeric_level)

    input_path = args.input
    output_path = args.output

    # Check if input file exists
    if not os.path.exists(input_path):
        logger.error(f"Input file '{input_path}' does not exist.")
        sys.exit(1)

    # Check if output file exists
    if os.path.exists(output_path) and not args.overwrite:
        logger.error(f"Output file '{output_path}' already exists. Use --overwrite to overwrite it.")
        sys.exit(1)

    # Perform the conversion
    convert_model(input_path, output_path)

if __name__ == "__main__":
    main()

File: ai/converter.go
package main

import (
	"bytes"
	"context"
	"encoding/gob"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"time"

	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
)

// ConverterConfig defines the configurations for both Python-based and native converters.
type ConverterConfig struct {
	PythonPath string        // Path to the Python interpreter
	ScriptPath string        // Path to converter.py
	LogDir     string        // Directory to store logs
	Timeout    time.Duration // Maximum time allowed for conversion
	NativeMode bool          // Use native conversion if true
}

// InitializeZapLogger initializes a Zap logger with a dynamic log file name.
func InitializeZapLogger(config ConverterConfig) (*zap.Logger, error) {
	// Ensure the log directory exists
	if err := os.MkdirAll(config.LogDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create log directory: %w", err)
	}

	// Generate a timestamp in ISO8601 format, replacing characters unsuitable for filenames
	timestamp := time.Now().UTC().Format("2006-01-02T15-04-05Z")
	logFileName := fmt.Sprintf("converter_log_%s.log", timestamp)
	logFilePath := filepath.Join(config.LogDir, logFileName)

	// Configure Zap logger
	cfg := zap.Config{
		Encoding:         "json", // Use JSON for structured logging
		Level:            zap.NewAtomicLevelAt(zap.InfoLevel),
		OutputPaths:      []string{logFilePath, "stdout"},
		ErrorOutputPaths: []string{"stderr"},
		EncoderConfig: zapcore.EncoderConfig{
			TimeKey:        "timestamp",
			LevelKey:       "level",
			NameKey:        "logger",
			CallerKey:      "caller",
			MessageKey:     "message",
			StacktraceKey:  "stacktrace",
			LineEnding:     zapcore.DefaultLineEnding,
			EncodeLevel:    zapcore.CapitalLevelEncoder,
			EncodeTime:     zapcore.ISO8601TimeEncoder, // ISO8601 format
			EncodeDuration: zapcore.StringDurationEncoder,
			EncodeCaller:   zapcore.ShortCallerEncoder,
		},
	}

	logger, err := cfg.Build()
	if err != nil {
		return nil, fmt.Errorf("failed to build Zap logger: %w", err)
	}

	return logger, nil
}

// ConvertModel invokes the Python converter script to convert AI models.
func ConvertModel(ctx context.Context, config ConverterConfig, logger *zap.Logger, inputPath, outputPath string, overwrite bool) error {
	if config.NativeMode {
		logger.Info("Using native converter")
		return NativeConvertModel(inputPath, outputPath, logger)
	}

	// Prepare command arguments
	args := []string{config.ScriptPath, inputPath, outputPath}
	if overwrite {
		args = append(args, "--overwrite")
	}

	// Create the command
	cmd := exec.CommandContext(ctx, config.PythonPath, args...)

	// Capture stdout and stderr
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	logger.Info("Starting Python-based model conversion", zap.String("input", inputPath), zap.String("output", outputPath))

	// Run the command
	if err := cmd.Start(); err != nil {
		logger.Error("Failed to start Python converter script", zap.Error(err))
		return fmt.Errorf("failed to start Python converter script: %w", err)
	}

	// Wait for the command to finish or context to timeout/cancel
	err := cmd.Wait()

	// Log stdout and stderr
	if stdout.Len() > 0 {
		logger.Debug("Python converter stdout", zap.String("output", stdout.String()))
	}
	if stderr.Len() > 0 {
		logger.Warn("Python converter stderr", zap.String("error_output", stderr.String()))
	}

	if err != nil {
		logger.Error("Model conversion failed", zap.Error(err))
		return fmt.Errorf("model conversion failed: %w", err)
	}

	logger.Info("Python-based model conversion completed successfully")
	return nil
}

// PyTorchModel represents a simplified PyTorch model structure.
type PyTorchModel struct {
	Parameters map[string][]float64
}

// GGMLModel represents a simplified GGML model structure.
type GGMLModel struct {
	Parameters map[string][]float64
}

// LoadPyTorchModel loads a simplified PyTorch `.pth` model.
func LoadPyTorchModel(inputPath string, logger *zap.Logger) (*PyTorchModel, error) {
	logger.Info("Loading PyTorch model", zap.String("input_path", inputPath))

	file, err := os.Open(inputPath)
	if err != nil {
		logger.Error("Failed to open PyTorch model file", zap.Error(err))
		return nil, err
	}
	defer file.Close()

	decoder := gob.NewDecoder(file)
	var model PyTorchModel
	if err := decoder.Decode(&model); err != nil {
		logger.Error("Failed to decode PyTorch model", zap.Error(err))
		return nil, err
	}

	logger.Info("PyTorch model loaded successfully", zap.Int("parameters_count", len(model.Parameters)))
	return &model, nil
}

// ConvertToGGML converts a PyTorch model to a GGML-compatible format.
func ConvertToGGML(pytorchModel *PyTorchModel, logger *zap.Logger) (*GGMLModel, error) {
	logger.Info("Converting PyTorch model to GGML format")

	ggmlModel := &GGMLModel{
		Parameters: make(map[string][]float64),
	}

	// Simplified conversion logic
	for name, params := range pytorchModel.Parameters {
		ggmlModel.Parameters[name] = params
		logger.Debug("Converted parameter", zap.String("name", name))
	}

	logger.Info("Conversion to GGML format completed", zap.Int("parameters_count", len(ggmlModel.Parameters)))
	return ggmlModel, nil
}

// SaveGGMLModel saves the GGML model to disk.
func SaveGGMLModel(ggmlModel *GGMLModel, outputPath string, logger *zap.Logger) error {
	logger.Info("Saving GGML model", zap.String("output_path", outputPath))

	file, err := os.Create(outputPath)
	if err != nil {
		logger.Error("Failed to create GGML model file", zap.Error(err))
		return err
	}
	defer file.Close()

	encoder := gob.NewEncoder(file)
	if err := encoder.Encode(ggmlModel); err != nil {
		logger.Error("Failed to encode GGML model", zap.Error(err))
		return err
	}

	logger.Info("GGML model saved successfully")
	return nil
}

// NativeConvertModel performs the entire conversion process using the native method.
func NativeConvertModel(inputPath, outputPath string, logger *zap.Logger) error {
	// Load PyTorch model
	pytorchModel, err := LoadPyTorchModel(inputPath, logger)
	if err != nil {
		return fmt.Errorf("failed to load PyTorch model: %w", err)
	}

	// Convert to GGML
	ggmlModel, err := ConvertToGGML(pytorchModel, logger)
	if err != nil {
		return fmt.Errorf("failed to convert to GGML: %w", err)
	}

	// Save GGML model
	if err := SaveGGMLModel(ggmlModel, outputPath, logger); err != nil {
		return fmt.Errorf("failed to save GGML model: %w", err)
	}

	return nil
}

func main() {
	// Define converter configuration
	config := ConverterConfig{
		PythonPath: "/usr/bin/python3", // Adjust path as necessary
		ScriptPath: filepath.Join("ghostshell", "converter", "converter.py"),
		LogDir:     filepath.Join("ghostshell", "logging"),
		Timeout:    5 * time.Minute, // Adjust timeout as necessary
		NativeMode: false,           // Set to true for native conversion
	}

	// Initialize Zap logger
	logger, err := InitializeZapLogger(config)
	if err != nil {
		fmt.Printf("Failed to initialize logger: %v\n", err)
		os.Exit(1)
	}
	defer logger.Sync()

	// Parse command-line arguments
	if len(os.Args) < 3 {
		logger.Error("Insufficient arguments provided. Usage: converter <input_path> <output_path> [--overwrite] [--native]")
		fmt.Println("Usage: converter <input_path> <output_path> [--overwrite] [--native]")
		os.Exit(1)
	}

	inputPath := os.Args[1]
	outputPath := os.Args[2]
	overwrite := false
	if len(os.Args) > 3 && os.Args[3] == "--overwrite" {
		overwrite = true
	}
	if len(os.Args) > 4 && os.Args[4] == "--native" {
		config.NativeMode = true
	}

	// Check if input file exists
	if _, err := os.Stat(inputPath); os.IsNotExist(err) {
		logger.Error("Input file does not exist", zap.String("input_path", inputPath))
		fmt.Printf("Error: Input file '%s' does not exist.\n", inputPath)
		os.Exit(1)
	}

	// Check if output file exists
	if _, err := os.Stat(outputPath); err == nil && !overwrite {
		logger.Error("Output file already exists", zap.String("output_path", outputPath))
		fmt.Printf("Error: Output file '%s' already exists. Use --overwrite to overwrite it.\n", outputPath)
		os.Exit(1)
	}

	// Create a context with timeout
	ctx, cancel := context.WithTimeout(context.Background(), config.Timeout)
	defer cancel()

	// Perform the conversion
	if err := ConvertModel(ctx, config, logger, inputPath, outputPath, overwrite); err != nil {
		logger.Fatal("Conversion process terminated with errors", zap.Error(err))
	}

	logger.Info("Model conversion completed successfully")
	fmt.Println("Model conversion completed successfully.")
}
